SilverSolver/assignment3/word_search.py

```diff
@@ -0,0 +1,49 @@
+from bisect import bisect_left
+from dictionary import Dictionary
+from typing import List, Tuple
+
+def empty_matrix(m: int, n: int, value: bool) -> List[List[bool]]:
+    # value should has immutable type!
+    result = []
+    for i in range(m):
+        result.append([value] * n)
+    return result
+
+def word_search(grid: List[List[str]], dictionary: Dictionary) -> List[str]: 
+    # -- Should we check grid correctness?
+    def neighbours(coords: Tuple[int]) -> Tuple[Tuple[int]]:
+        a, b = coords
+        neigh_coords = [(a + i, b + j) for i in [-1, 1, 0] for j in [-1, 1, 0]]
+        del neigh_coords[-1]
+        return filter(lambda x: x[0] >= 0 and x[1] >= 0 and x[0] < m and x[1] < n, neigh_coords)
```

Comment:
2 notes:
- it is recommended to use a list comprehension instead of creating a lambda. Last time I checked, it was also faster.
- Python supports interval checks as boolean expression: this is correct: 0 <= a < max
- Python supports unpacking. Ex: [(element[1], element[0]) for element in my_list] is the same as [(y, x) for x, y in my_list]

You can write this:
  return [(x, y) for x, y in neigh_coords if 0 <= 0 x < m and 0 <= y < m]